compiler/rustc_trait_selection/src/solve/eval_ctxt.rs

```diff
@@ -173,6 +173,7 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
             };
             let (ref infcx, goal, var_values) = tcx
                 .infer_ctxt()
+                .with_new_solver()
```

Comment:
would be cool to land this as yet another `-Z` flag, e.g. `-Ztrait-solver=next-coherence` or whatever